crates/matrix-sdk/src/sliding_sync/list/mod.rs

```diff
@@ -300,6 +313,23 @@ pub(super) struct SlidingSyncListInner {
 }
 
 impl SlidingSyncListInner {
+    /// Change the sync-mode.
+    ///
+    /// This will change the sync-mode but also the request generator.
+    ///
+    /// [`Self::ranges`] and [`Self::state`] will be updated when the next
+    /// request will sent and a response will be received. The
+    /// [`Self::maximum_number_of_rooms`] won't change.
+    pub fn set_sync_mode(&self, sync_mode: SlidingSyncMode) {
+        // Acquire both locks before modifying the values their hold.
+        let mut sync_mode_lock = self.sync_mode.write().unwrap();
```

Comment:
`sync_mode` is checked whenever we try to call `set_ranges`; i see that calling `set_ranges` isn't allowed when not in selective mode. So that means I could theoretically do the following sequence of calls on a list:

- `set_sync_mode(Selective)`
- `set_ranges(whatever_value_i_want)`
- `set_sync_mode(Growing)`

And that way, I would circumvent the sync mode check in `set_ranges`. Is that expected/allowed, or should this method become fallible if ranges have been set, or reset the ranges with respect to the target `sync_mode` value? (if not allowed, can you tweak/add tests for this please?)
If that's not a problem, can you comment about it in this function's doc comment?

Reply:
As soon as the sync mode will be set to `Growing` or `Paging`, the ranges will be calculated automatically by the request generator (`SlidingSyncListRequestGenerator`), in `Selective` mode, the ranges provided by the user are used by the request generator.

When the sync mode is set, a fresh new request generator is created, it's similar to a reset. And then, when the response is received, the ranges are updated (in `SlidingSyncListInner::update_request_generator_state`). So whatever you've set in `Selective` mode, it will be overwritten by the update of the request generator. The request generator is the conductor whatever the sync mode is.

Maybe I should clarify the role of the request generator in the documentation.

Reply:
Oh I see, thanks for the explanation. In this case, it sounds like there's state duplication: ranges is in both the generator and the inner sliding sync. In particular, it seems that `range` is only meaningful when the list mod is `Selective`. Would it make sense to keep the range in the request generator, instead, so that it's set only in `Selective` mode, and not duplicated anymore?

Then, I would also change the API so that `SlidingSync::set_ranges` and all are not on the `SlidingSync` anymore, but on the request generator itself, and forwarded via a mutator object `SlidingSync::ranges_mut() -> &mut RangeParameters`, with `RangeParameters` that has some methods `set` / `add` etc.

Maybe that's a big ask, and it could be done as a follow-up; I'd be happy to try that, actually.